velox/functions/prestosql/Reduce.cpp

```diff
@@ -67,39 +67,47 @@ class ReduceFunction : public exec::VectorFunction {
   void apply(
       const SelectivityVector& rows,
       std::vector<VectorPtr>& args,
-      const TypePtr& /* outputType */,
+      const TypePtr& outputType,
       exec::EvalCtx& context,
       VectorPtr& result) const override {
     VELOX_CHECK_EQ(args.size(), 4);
-
+    // Identify the rows need to be computed.
+    exec::LocalSelectivityVector nonNullRowsHolder(*context.execCtx());
```

Comment:
You can calculate the nonNullRows after creating the flatArray on line 90 using the flatArray instead of the decoded vector. Also, please maintain the use of exec::LocalDecodedVector in the original code instead of using the DecodedVector class here directly.

Reply:
updated.